RFC and Prototype for Authentication Integration

shiroyuki opened the pull request:
(Related to #58)

RFC: https://allspark.dev.data.humancellatlas.org/dcp-ops/docs/wikis/RFC:-Authentication-Flow

shiroyuki commented on src/azul/security/verifier.py:
```diff
@@ -0,0 +1,98 @@
+        raise InvalidTokenError('missing information')
+    except PyJWTError:
+        logger.warning(f"Detected a JWT with INVALID SIGNATURE (given: '{token}').", exc_info=True)
+        raise UnverifiableTokenError(token)
```
I agree that we can just raise `InvalidTokenError` here. But I have some idea to utilize this error.

Initially when I implement these exceptions, I thought of how the caller code or maintainers should respond according to a type of error.

| Exception | Interpretation |
| --- | --- |
| `InvalidTokenError` | The token is broken. No security risk. |
| `NonDecodableTokenError` | The token is not decodable. Probably random token. Low security risk |
| `UnverifiableTokenError` | The token is decodable but fails the verification. Expired token? Tampered token? |

So, in the last case, I wonder if we should just log the messages in the warning level and treat as normal invalid token error. Or later on, we should also notify to the DevOps team.

What do you think, @hannes-ucsc?

shiroyuki commented on src/azul/security/verifier.py:
```diff
@@ -0,0 +1,98 @@
+
+
+def convert_base64_string_to_int(value:str) -> int:
+    return int.from_bytes(urlsafe_b64decode(value + "==="), byteorder="big")
```
But I'll fix that.

shiroyuki commented:
I would agree with you that option 2 makes the code more manageable and portable. I'll wait for more feedback from @NoopDog.

hannes-ucsc commented:
Can you resolve the conflicts. Also where is the wheel you mentioned earlier?

shiroyuki commented:
@hannes-ucsc The unpacked wheels have been pushed to `lambdas/service/vendor`. I'm working on the merge conflict.

shiroyuki commented:
@hannes-ucsc Merge conflict resolved.

shiroyuki commented:
Switched back to "in progress" temporarily to add missing tests."

shiroyuki commented:
For clarification, the review request is for the RFC on the GitLab.

shiroyuki commented:
**Summary of notable changes up to d49b8ba**

* As patching (with `unittest.mock`) doesn't work when for the endpoint tests,
  * the base class for app test cases will add three endpoints to provide JWKS URI, public keys and generate for-testing-only JWTs to make the app a temporary authentication broker,
  * when the tests run, the private and public keys will be automatically generated once where the key files (actually PEM files) are listed in the `.gitignore` file,
  * the verifier always accept issuers from `http://127.0.0.1:\d{1,5}`.

What have not been addressed (still in progress)?

* Weird padding in `azul.security.verifier.convert_base64_string_to_int`

Please note that these are files added only for deployment.

* `lambdas/service/vendor/.libs_cffi_backend/libffi-d78936b1.so.6.0.4`
* `lambdas/service/vendor/_cffi_backend.cpython-36m-x86_64-linux-gnu.so`
* `lambdas/service/vendor/cffi*`
* `lambdas/service/vendor/cryptography*`

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -769,13 +806,67 @@ def files_proxy(uuid):
         dss_response.raise_for_status()
 
 
+def self_url(endpoint_path=None):
```
This method already exists on `develop`.

shiroyuki replied:
The duplicate has been removed.

hannes-ucsc commented on .gitignore:
```diff
@@ -7,6 +7,10 @@
 #
 # - No leading slash means the pattern matches at any depth.
 
+# OpenSSL Private-Public Keys for Testing
```
```suggestion
# OpenSSL private/public keys for testing authentication
```

But can we we delete these during test clean-up?

hannes-ucsc commented:
So now there are two reviews. GitHub must have troubles today.

shiroyuki commented:
Summary of Changes:
* Replaced uncompressed wheel/source files with the compressed version.
* Updated `lambdas/service/Makefile` to uncompress the files before `make deploy`.
* The test keys are created and destroy on test app's start and end respectively. (i.e., the keys will be clean up after the test.)
* Resolved conflicts.

hannes-ucsc commented on test/app_test_case.py:
```diff
@@ -1,17 +1,26 @@
 # noinspection PyPackageRequirements
 from chalice.config import Config as ChaliceConfig
-from azul import config
+from azul import config as azul_config
```
I want this to be`config` consistently across the code base. If some other name conflicts with that, I want that other thing to be renamed.

shiroyuki replied:
ok

shiroyuki replied:
The alias has been removed.

shiroyuki commented:
@hannes-ucsc Sorry, I forgot to check those wheel files in.

Summary of changes (as of 29f87045461dacf246841ef66ac3127474b5ab76):
* Simplified the deployment process (that part that uncompresses the wheel files).
* Added the missing wheel files.
* Moved key generation/removal to setUpClass and tearDownClass to speed up the tests.

hannes-ucsc commented on .gitignore:
```diff
@@ -7,6 +7,10 @@
 #
 # - No leading slash means the pattern matches at any depth.
 
+# OpenSSL Private-Public Keys for testing authentication
```
Why is this STILL here?

hannes-ucsc commented on src/azul/__init__.py:
```diff
@@ -309,6 +309,17 @@ def cart_item_state_machine_name(self):
 
+    @property
+    def access_token_issuer(self):
+        return "https://humancellatlas.auth0.com/"
```
```suggestion
        return "https://humancellatlas.auth0.com"
```

If at all possible. We use base URLs that don't end in slash because …

hannes-ucsc commented on src/azul/__init__.py:
```diff
@@ -309,6 +309,17 @@ def cart_item_state_machine_name(self):
+    def access_token_audience_list(self):
+        return [
+            f"https://{self.deployment_stage}.data.humancellatlas.org/",
+            f"{self.access_token_issuer}userinfo"
```
… it makes this look nicer:

```suggestion
            f"{self.access_token_issuer}/userinfo"
```

hannes-ucsc commented on src/azul/security/auth.py:
```diff
@@ -0,0 +1,102 @@
+
+logger = getLogger(__name__)
+
+fusillade_callback_url = f'{config.service_endpoint()}/auth/callback'
```
```suggestion
fusillade_callback_url = {config.service_endpoint() + '/auth/callback'
```

hannes-ucsc replied:
Reads more clearly due to the whitespace around +

shiroyuki replied:
Oh... I see. This will be changed in the next commit.

hannes-ucsc commented on src/azul/security/auth.py:
```diff
@@ -0,0 +1,102 @@
+
+
+def get_fusillade_hostname():
+    # This can be moved to config.
```
Please do that.

hannes-ucsc commented on src/azul/security/auth.py:
```diff
@@ -0,0 +1,102 @@
+    return f'{fusillade_subdomain}.data.humancellatlas.org'
+
+
+def get_fusillade_stage() -> str:
```
I don't like expressing this mapping from deployment stage to Fusillade stage as code. We use `deployments/*/environment` for expressing differences between deployments.

hannes-ucsc commented on src/azul/security/auth.py:
```diff
@@ -0,0 +1,102 @@
+        assert authorization_token.startswith(bearer_token_prefix), "not_bearer_token"
+        access_token = authorization_token[len(bearer_token_prefix):]
+        assert access_token, "missing_bearer_token"
+    except AssertionError as e:
```
This is misuse of assertions.

hannes-ucsc commented on src/azul/security/verifier.py:
```diff
@@ -0,0 +1,97 @@
+
+from azul import config
+
+logger = getLogger(__name__)
```
```suggestion
logger = logging.getLogger(__name__)
```

shiroyuki commented:
@hannes-ucsc I think the problem might be due to Python 3.7, which I install via `brew`. But I'll investigate further.

hannes-ucsc commented:
Indeed, it is caused by Python 3.7. @danielsotirhos ran into a similar issue.

shiroyuki commented:
@hannes-ucsc In that case, I think we don't need these wheels.

shiroyuki commented:
I'll remove those wheels.

hannes-ucsc commented on common.mk:
```diff
@@ -34,3 +34,7 @@ endif
 # template creation. This is what the fake .FORCE target does. It still is necessary to declare a target's dependency on
 # a template to ensure correct ordering.
+
+vendor/%.dist-info: $(AZUL_HOME)/wheels/%-*-*-*.whl
```
Yes, please remove.

shiroyuki commented:
This pattern rule has been removed.
``` Makefile
vendor/%.dist-info: $(AZUL_HOME)/wheels/%-*-*-*.whl
	unzip $< -d vendor
	test -x $@
```

shiroyuki commented:
The wheels are now removed.

hannes-ucsc commented on src/azul/security/verifier.py:
```diff
@@ -0,0 +1,111 @@
+gserviceaccount_domain = "iam.gserviceaccount.com"
+
+# Shared HTTP session
+session = requests.Session()
```
No side-effects at module loading/import time, please. Please fix in a follow-up PR.

hannes-ucsc commented on environment:
```diff
@@ -226,6 +229,8 @@ _set AZUL_CART_API_IP_WHITELIST  # TODO: Remove when authentication is added
 
 _set azul_dss_query_prefix
 
+_set AZUL_FUSILLADE_HOSTNAME "auth.dev.data.humancellatlas.org"
```
Other services endpoints are configured via a "endpoint" variable e.g., a base URL to which we append. This would allow for more flexible configuration.

hannes-ucsc commented:
Sorry, I found merge errors after merging and backed this out again. Please file a new PR with 

- `AZUL_FUSSILADE_ENDPOINT` instead of AZUL_FUSILLADE_HOSTNAME
- no side effect at module import time
- and the merge error fixed (I'll point it out) in a review comment

hannes-ucsc commented on test/app_test_case.py:
```diff
@@ -65,6 +163,10 @@ def base_url(self):
     def setUpClass(cls):
         super().setUpClass()
+        generate_test_keys()
+        cls._path_to_app = os.path.join(config.project_root, 'lambdas', cls.lambda_name())
```
Merge error, please remove.

hannes-ucsc commented on test/app_test_case.py:
```diff
@@ -65,6 +163,10 @@ def base_url(self):
         super().setUpClass()
+        generate_test_keys()
+        cls._path_to_app = os.path.join(config.project_root, 'lambdas', cls.lambda_name())
+        sys.path.append(cls._path_to_app)
```
Merge error, please remove.

hannes-ucsc commented on test/app_test_case.py:
```diff
@@ -74,6 +176,8 @@ def setUpClass(cls):
     @classmethod
     def tearDownClass(cls):
         cls.app_module = None
+        sys.path.remove(cls._path_to_app)
```
Merge error, please remove.

shiroyuki commented:
Got it. I'll look into it now.
